clarity/contracts/helpers/token-lock.clar

```diff
@@ -0,0 +1,171 @@
+(define-map approved-workers principal bool)
+(define-map token-schedule 
+    {
+        token: principal,
+        address: principal,
+        block-height: uint
+    }
+    uint
+)
+
+(define-read-only (get-contract-owner)
+  (ok (var-get contract-owner))
+)
+
+(define-public (set-contract-owner (owner principal))
+  (begin
+    (try! (check-is-owner))
+    (ok (var-set contract-owner owner))
+  )
+)
+
+(define-private (check-is-owner)
+  (ok (asserts! (is-eq tx-sender (var-get contract-owner)) ERR-NOT-AUTHORIZED))
+)
+
+(define-public (set-approved-token (token principal) (approved bool))
+	(begin
+		(try! (check-is-owner))
+		(ok (map-set approved-tokens token approved))
+	)
+)
+
+(define-private (check-is-approved-token (token principal))
+  (ok (asserts! (default-to false (map-get? approved-tokens token)) ERR-TOKEN-NOT-AUTHORIZED))
+)
+
+(define-public (set-approved-worker (worker principal) (approved bool))
+	(begin
+		(try! (check-is-owner))
+		(ok (map-set approved-workers worker approved))
+	)
+)
+
+(define-private (check-is-approved-worker (worker principal))
+  (ok (asserts! (default-to false (map-get? approved-workers worker)) ERR-WORKER-NOT-AUTHORIZED))
+)
+
+(define-read-only (get-tokens-to-vest-or-default (token principal))
+    (default-to u0 (map-get? tokens-to-vest token))
+)
+
+(define-public (set-token-schedule (token principal) (address principal) (height uint) (amount uint))
+    (begin 
+        (try! (check-is-owner))
+        (try! (check-is-approved-token token))
+        (map-set token-schedule { token: token, address: address, block-height: height } amount)
+        (ok (map-set tokens-to-vest token (+ (get-tokens-to-vest-or-default token) amount)))    
+    )
+)
+
+(define-private (set-token-schedule-iter (item { token: principal, address: principal, block-height: uint, amount: uint }))
+    (set-token-schedule (get token item) (get address item) (get block-height item) (get amount item))
+)
+
+(define-public (set-token-schedule-many (items (list 200 { token: principal, address: principal, block-height: uint, amount: uint })))
+    (ok (map set-token-schedule-iter items))
+)
+
+(define-read-only (get-token-schedule-or-fail (token principal) (address principal) (height uint))
+    (ok (unwrap! (map-get? token-schedule { token: token, address: address, block-height: height }) ERR-SCHEDULE-NOT-FOUND))
+)
+
+(define-private (get-tokens-iter (token-trait <ft-trait>) (address principal) (height uint))
+    (let 
+        (
+            (token (contract-of token-trait))
+            (amount (try! (get-token-schedule-or-fail token address height)))
+        )
+        (asserts! (> block-height height) ERR-BLOCK-HEIGHT-NOT-REACHED)
+        (map-set token-schedule { token: token, address: address, block-height: height } u0)
+        (map-set tokens-to-vest token (- (get-tokens-to-vest-or-default token) amount))
+        (as-contract (try! (contract-call? token-trait transfer-fixed amount tx-sender address none)))
```

Comment:
Vesting may fail if the `token-lock` contract has insufficient tokens. How will the contract receive the tokens? On deploy? (And an off-by-one error may lock tokens until a worker fixes it manually.)